### DISCUSSION TITLE
Allow the diff3 conflict style in web console conflict resolver.

### DISCUSSION THREAD

**Discussion by JacksonBaileyHM** [community/community#5011] [Other Feature Feedback, Questions, & Ideas]
To be frank, the default conflict style can be utterly useless in many scenarios. When two people change the same line of code it's often needed to know how both of those differ from the original to make sense of how to merge it. The `diff3` style allows this by showing your code, their code, and the original.

For more information about it, click [here](https://git-scm.com/book/en/v2/Git-Tools-Advanced-Merging) but I've quoted the relevant information below.

> Let’s explore a couple of tools that you now have at your disposal to determine how this conflict came to be. Perhaps it’s not obvious how exactly you should fix this conflict. You need more context.
> 
> One helpful tool is `git checkout` with the `--conflict` option. This will re-checkout the file again and replace the merge conflict markers. This can be useful if you want to reset the markers and try to resolve them again.
> 
> You can pass `--conflict` either `diff3` or `merge` (which is the default). If you pass it `diff3`, Git will use a slightly different version of conflict markers, not only giving you the “ours” and “theirs” versions, but also the “base” version inline to give you more context.
> 
> ```bash
> $ git checkout --conflict=diff3 hello.rb
> ```
> 
> Once we run that, the file will look like this instead:
> 
> ```
> #! /usr/bin/env ruby
> 
> def hello
> <<<<<<< ours
>   puts 'hola world'
> ||||||| base
>   puts 'hello world'
> =======
>   puts 'hello mundo'
> >>>>>>> theirs
> end
> 
> hello()
> ```
> 
> If you like this format, you can set it as the default for future merge conflicts by setting the `merge.conflictstyle` setting to `diff3`.
> 
> ```bash
> $ git config --global merge.conflictstyle diff3
> ```

**Comment by gnprice** [comment]
Heartily agreed. To expand a bit on the description, in case it's helpful: it's quite common that the information provided by the default conflict style leaves it *completely impossible* to correctly resolve the conflict. You end up having to look up the history of the changes you're merging, or else just guessing.

With the `diff3` style, by contrast, almost all merge conflicts can be reliably resolved with just the information in the conflict hunk. That makes it possible to go through and resolve a bunch of conflicts much more quickly, as well as with many fewer errors.

Concretely, taking the same example as in the description above, the default conflict style looks like this:
```
<<<<<<< HEAD
  puts 'hola world'
=======
  puts 'hello mundo'
>>>>>>> some-feature
```

If you look at that, you can pretty well figure that the right merged version will be either `puts 'hola mundo'` or `puts 'hello world'`. But which one is it?

It could really just as well be either one. If the line originally said `hello world`, then one branch changed `hello` to `hola` and the other changed `world` to `mundo`, so both changes combined should make it `hola mundo`. … But the original could equally have been `hola mundo`, and in that case one branch changed `mundo` to `world` while the other changed `hola` to `hello`, which would mean the result of both changes is `hello world`. The only way to disambiguate is to go sort through the history to work out which of those is what happened.

And sorting through the history like that should really be the merge engine's job. In fact it *is* the merge engine's job -- it already has to do that in order to get as far as it does with merging all the non-conflicting changes in the file. All it needs to do is include those historical findings in the conflict hunks it hands to us… and that's exactly what `diff3` does.

In our example, the conflict hunk might look (as the quote in the OP says) like:
```
<<<<<<< ours
  puts 'hola world'
||||||| base
  puts 'hello world'
=======
  puts 'hello mundo'
>>>>>>> theirs
```
Then we know the old version was `hello world`, so the changes were from that to `hola` and `mundo`, and the result of both changes together is `hola mundo`.

---

In my view, for these reasons `merge.conflictstyle=diff3` should have become the default in Git itself long ago. But independently of that, GitHub has a choice of what style to use in its own web UI, and it should use `diff3` there for the same reasons.

**Reply by Bqrtooo85** [reply]
Ggg

**Comment by acrefoot** [comment]
Our team is finding this to be a problem too, and increasingly we rely on the web console tools out of habit. 

Here's our use case for relying on the web console:
Rebasing using the web console instead of local git CLI seems to maintain a PR review approval even when "Dismiss stale pull request approvals when new commits are pushed" is turned on. (This is important when trying to merge a series of PRs on a fast-moving codebase.)

Being able to change the conflict style would really help make the web console tools first class!

**Comment by Jacob-Stevens-Haas** [comment]
Another thing that's important here is when a PR author needs help resolving a merge conflict, the solution is to commit the conflict so library maintainers can comment in review.

I want to add my use case and the general "workaround" in order to show why this is important.  I maintain a research package to which other postdocs and Ph.D. students contribute capabilities occasionally.  Given their reasonable desire to publish research before sharing the capability, they tend to have a branch that is six months to a year stale by the time they PR.  Most of the conflicts would have been reasonably resolvable by the PR author, but if they do so in the web UI, we would lose the diff3.  So the solution is to check out a new branch locally, merge the diff3 conflicts, and PR _that_ branch.

e.g. see [this PR](https://github.com/dynamicslab/pysindy/pull/459/)

**Comment by towerofnix** [comment]
Hi, this is a big-deal request on my part too! I review pull requests for hsmusic-data all the time, where changes are super frequently to lines of data (rather than code) that have been differently revised since the pull request opened. I basically ALWAYS have to view the common base to meaningfully interpret—and meaningfully *merge*—the changes. diff3 is a lifesaver and it's really sad not to have it on GitHub.
